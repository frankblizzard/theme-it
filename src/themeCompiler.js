import {
  camelToKebab,
  doubleQuoteRgx,
  commaRgx,
  rightParenSemicolonRgx,
  leadingAndTrailingParenRgx,
  colonSpaceLeftParenRgx
} from "./utils";

const formatOutput = output => JSON.stringify(output, null, 2);

const parseNestedPrefix = (prefix) => prefix ? `${prefix}-` : "";

const isNested = value => typeof value === "object";

/**
 * Parses the Themeit CSS Variables "AST" into valid CSS class definitions.
 * @param {Object} obj - Parsed Themeit CSS Variables Object
 */
function objectToCssString(obj) {
  return formatOutput(obj)
    .replace(doubleQuoteRgx, "")
    .replace(leadingAndTrailingParenRgx, "")
    .replace(commaRgx, ";")
    .replace(rightParenSemicolonRgx, "}")
    .replace(colonSpaceLeftParenRgx, " {");
}

/**
 * Recursive function that traverses the passed in object tree/JSON and creates a complete CSS variable theme structure from it. 
 * You can define styles the same way you would in react or a similar library, and theme-it will generate a complete CSS Variable definition for you
 * that can be used for theming.
 * @param {Object} themeObj - The theme object, such as the theme configuration object from a CSS in JS library.
 * @param {Object} result - The parsed and generated "AST" containing information about the css variables and their values. 
 * @param {String} prefix - Stateful buildup of CSS variable names when traversing a nested object.  
 */
function traverse(themeObj, result, prefix) {
  for (let cssVariableName in themeObj) {
    const cssVariableValue = themeObj[cssVariableName];
    const cssVariableKey = parseNestedPrefix(prefix) + cssVariableName;

    if (isNested(cssVariableValue)) {
      traverse(cssVariableValue, result, cssVariableKey);
    } else {
      result[`--${camelToKebab(cssVariableKey)}`] = cssVariableValue;
    }
  }
}

/**
 * Generates the CSS variable theme by passing in your theme styles as an object. Calls traverse() to do the conversion. 
 * @param {Object} themeConfig - Nested or Flat Theme configuration object
 */
const generateTheme = (themeConfig) => {
  const generated = {};
  for (let themeName in themeConfig) {
    const themeClass = "." + camelToKebab(themeName);
    generated[themeClass] = {};
    
    traverse(themeConfig[themeName], generated[themeClass]);
  }

  return generated;
};

/**
 * Generates a valid CSS definition as a string from the passed in theme configuration object.
 * @param {Object} themeConfig - theme configuration object
 * @param {Boolean} printCSS - print the generated CSS to the console for copy + paste or debugging.
 */
function generateCSSFromThemeObj(themeConfig, printCSS) {
  const cssVariableTree = generateTheme(themeConfig); 
  const cssRuleDefinition = objectToCssString(cssVariableTree);

  if (printCSS) {
    console.log(cssRuleDefinition);
  }

  return cssRuleDefinition;
}



export { generateTheme, generateCSSFromThemeObj };
