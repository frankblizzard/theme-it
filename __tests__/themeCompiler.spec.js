import { generateTheme, generateCSSFromThemeObj } from "themeCompiler";
import {
  testThemeConfig,
  generatedTheme,
  nestedThemeConfig,
  nestedGeneratedTheme
} from "./testData";

describe("Themeit theme compiler tests", () => {
  it("Should generate the correct CSS class structure when the themeConfig is passed in", () => {
    const generated = generateTheme(testThemeConfig);
    expect(generated).toEqual(generatedTheme);
  });

  it("Should generate the correct CSS class structure when the nestedThemeConfig is passed in", () => {
    const generated = generateTheme(nestedThemeConfig);
    expect(generated).toEqual(nestedGeneratedTheme);
  });

  it("Should generate the correct valid CSS when when the themeConfig is passed in", () => {
    expect(generateCSSFromThemeObj(testThemeConfig, true)).toMatchSnapshot();
  });

  it("Should generate the correct valid CSS when when the nestedThemeConfig is passed in", () => {
    expect(generateCSSFromThemeObj(nestedThemeConfig, true)).toMatchSnapshot();
  });
});
