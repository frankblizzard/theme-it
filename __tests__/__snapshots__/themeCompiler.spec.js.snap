// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Themeit theme compiler tests Should generate the correct valid CSS when when the nestedThemeConfig is passed in 1`] = `
"
  .dark-theme {
    --primary-color: black;
    --secondary-color: papayawhip;
    --global-border: 2px;
    --button-primary-color: blue;
    --input-secondary-color: papayawhip;
    --input-radio-color: purple;
    --input-text-border: 5px
  }
  .light-theme {
    --primary-color: gray;
    --secondary-color: purple;
    --global-border: 2px;
    --button-primary-color: blue;
    --input-secondary-color: papayawhip;
    --input-radio-color: purple;
    --input-text-border: 5px
  }
"
`;

exports[`Themeit theme compiler tests Should generate the correct valid CSS when when the themeConfig is passed in 1`] = `
"
  .dark-theme {
    --primary-color: black;
    --secondary-color: papayawhip;
    --global-border: 2px
  }
  .light-theme {
    --primary-color: gray;
    --secondary-color: purple;
    --global-border: 2px
  }
"
`;
